h5py/_hl/group.py

```diff
@@ -499,7 +499,10 @@ def copy(self, source, dest, name=None,
                     dest_path = name
                 else:
                     # copy source into dest group: dest_name/source_name
-                    dest_path = pp.basename(h5i.get_name(source[source_path].id))
+                    if isinstance(source, dataset.Dataset):
+                        dest_path = pp.basename(h5i.get_name(dest.id))
```

Comment:
I think this should use `source.id` rather than `dest.id`. That should fix the issue you noticed when writing the tests.